compiler/controllers.py

```diff
@@ -0,0 +1,281 @@
+    store = Store.current_session()
+    try:
+        stat = store.get_status(source_id, checksum, output_format)
+        logger.debug('Got status from store: %s', stat)
+        return stat
+    except DoesNotExist as e:
+        logger.debug('No such compilation: %s', e)
+    # except Exception as e:
+    #     logger.debug('No such compilation: %s', e)
+    return None
+
+
+def _redirect_to_status(source_id: str, checksum: str, output_format: Format,
+                        code: int = status.SEE_OTHER) -> Response:
+    """Redirect to the status endpoint."""
+    location = url_for('api.get_status', source_id=source_id,
+                       checksum=checksum, output_format=output_format.value)
+    return {}, code, {'Location': location}
+
+
+def service_status(*args: Any, **kwargs: Any) -> Response:
+    """Exercise dependencies and verify operational status."""
+    fm = filemanager.FileManager.current_session()
+    store = Store.current_session()
+    response_data = {}
+    response_data['store'] = store.is_available()
+    response_data['compiler'] = compiler.is_available()
+    response_data['filemanager'] = fm.is_available()
+    if not all(response_data.values()):
+        return response_data, status.SERVICE_UNAVAILABLE, {}
+    return response_data, status.OK, {}
+
+
+def compile(request_data: MultiDict, token: str, session: Session,
+            is_authorized: Callable = lambda task: True) -> Response:
+    """
+    Start compilation of an upload workspace.
+
+    Parameters
+    ----------
+    request_data : :class:`.MultiDict`
+        Data payload from the request.
+    token : str
+        Auth token to be used for subrequests (e.g. to file management
+        service).
+
+    Returns
+    -------
+    dict
+        Response data.
+    int
+        HTTP status code.
+    dict
+        Headers to add to response.
+
+    """
+    if 'output_format' in request_data:
+        output_format = Format(request_data['output_format'])
+    else:
+        output_format = Format.PDF
+    source_id = request_data.get('source_id', None)
+    checksum = request_data.get('checksum', None)
+    if source_id is None:
+        logger.debug('Missing required parameter: source_id')
+        raise BadRequest('Missing required parameter: source_id')
+    if checksum is None:
+        logger.debug('Missing required parameter: checksum')
+        raise BadRequest('Missing required parameter: checksum')
+
+    # We don't want to compile the same source package twice.
+    force = request_data.get('force', False)
+
+    # Support label and link for PS/PDF Stamping
+    # Test
+    stamp_label: Optional[str] = request_data.get('stamp_label', None)
+    stamp_link: Optional[str] = request_data.get('stamp_link', None)
+
+    logger.debug('%s: request compilation with %s', __name__, request_data)
+    if not force:
+        info = _status_from_store(source_id, checksum, output_format)
+        if info is not None:
```

Comment:
Should `NotFound('No such resource')` be raised if `info` is `None`?

Reply:
If ``info`` is ``None``, we're good to go and can can proceed with compilation in the non-force case. Added a comment in b403024 to clarify:

```
    # Unless we are forcing recompilation, we do not want to compile the same 
    # source twice. So we check our storage for a compilation (successful or 
    # not) corresponding to the requested source package.
```